wheatley/tower.py

```diff
@@ -216,60 +212,57 @@ def _on_user_leave(self, data):
         # Unassign all instances of that user
         for bell, user in self._assigned_users.items():
             if user == user_id_that_left:
-                self._assigned_users[bell] = None
+                del self._assigned_users[bell]
 
                 bells_unassigned.append(bell)
 
         self.logger.info(
             f"RECEIVED: User #{user_id_that_left}:'{user_name_that_left}' left from bells {bells_unassigned}."
         )
 
-    def _on_user_entered(self, data):
+    def _on_user_entered(self, data: JSON) -> None:
         """ Called when the server receives a uew user so we can update our user list. """
         # Add the new user to the user list, so we can match up their ID with their username
         self._user_name_map[data['user_id']] = data['username']
 
-    def _on_user_list(self, user_list):
+    def _on_user_list(self, user_list: JSON) -> None:
         """ Called when the server broadcasts a user list when Wheatley joins a tower. """
         for user in user_list['user_list']:
             self._user_name_map[user['user_id']] = user['username']
 
-    def _join_tower(self):
+    def _join_tower(self) -> None:
         """ Joins the tower as an anonymous user. """
         self._emit(
             "c_join",
             {"anonymous_user": True, "tower_id": self.tower_id},
             f"Joining tower {self.tower_id}"
         )
 
-    def _request_global_state(self):
+    def _request_global_state(self) -> None:
         """ Send a request to the server to get the current state of the tower. """
         self._emit('c_request_global_state', {"tower_id": self.tower_id}, "Request state")
 
-    def _emit(self, event: str, data, message: str):
-        """ Emit a socket-io signal. """
-        if self._socket_io_client is None:
-            raise Exception("Not Connected")
-
-        self._socket_io_client.emit(event, data)
-
-        if message:
-            self.logger.info(f"EMIT: {message}")
-
-    def _on_bell_rung(self, data):
+    def _on_bell_rung(self, data: JSON) -> None:
         """ Callback called when the client receives a signal that a bell has been rung. """
         self._update_bell_state(data["global_bell_state"])
 
         who_rang = Bell.from_number(data["who_rang"])
+        # Only run the callbacks if the bells exist
         for bell_ring_callback in self.invoke_on_bell_rung:
-            bell_ring_callback(who_rang, self.get_stroke(who_rang))
-
-    def _update_bell_state(self, bell_state):
+            new_stroke = self.get_stroke(who_rang)
+            if new_stroke is None:
+                self.logger.warning(
+                    f"Bell {data['who_rang']} rang, but the tower only has {self.number_of_bells} bells."
+                )
+            else:
+                bell_ring_callback(who_rang, new_stroke)
+
+    def _update_bell_state(self, bell_state: List[Stroke]) -> None:
```

Comment:
I think this is still receiving the raw bools from JSON instead of the new Stroke class

Reply:
Gooood spot